docs/src/main/asciidoc/tls-registry-reference.adoc

```diff
@@ -0,0 +1,518 @@
+
+To access the TLS configuration, inject the `TlsConfigurationRegistry` bean and gets the TLS configuration by name (or the default one):
+
+[source,java]
+----
+ @Inject
+ TlsConfigurationRegistry certificates;
+// ...
+TlsConfiguration def = certificates.getDefault().orElseThrow();
+TlsConfiguration named = certificates.get("name").orElseThrow();
+//...
+----
+
+The `TlsConfiguration` object contains the key stores, trust stores, cipher suites, protocols, and other properties.
+It also provides a way to create an `SSLContext` from the configuration.
+
+As Vert.x is omnipresent in Quarkus, you can also use the `TlsConfiguration` object to configure the Vert.x client or server such as `KeyCertOptions`, `TrustOptions`, and so on.
+
+== Registering a certificate from an extension
+
+This section is only for extension developers.
+An extension can register a certificate in the TLS registry.
+This is useful when an extension needs to provide a certificate to the application or provides a different format.
+
+To achieve this, the extension _processor_ must produce a `TlsCertificateBuildItem`.
+A `TlsCertificateBuildItem` is composed of a name and a `CertificateSupplier`.
+
+[source,java]
+----
+TlsCertificateBuildItem item = new TlsCertificateBuildItem("named",
+    new MyCertificateSupplier());
+----
+
+The certificate supplier is a runtime object that is generally retrieved using a recorder method.
+Here is an example of a certificate supplier:
+
+[source,java]
+----
+public class MyCertificateSupplier implements Supplier<TlsConfiguration> {
+
+        @Override
+        public TlsConfiguration get() {
+            try {
+                KeyStore ks = KeyStore.getInstance("PKCS12");
+                ks.load(getClass().getResourceAsStream("target/certs/test-registration-keystore.p12"),
+                        "password".toCharArray());
+                KeyStore ts = KeyStore.getInstance("PKCS12");
+                ts.load(getClass().getResourceAsStream("target/certs/test-registration-truststore.p12"),
+                        "password".toCharArray());
+                return new BaseTlsConfiguration() {
+                    @Override
+                    public KeyStore getKeyStore() {
+                        return ks;
+                    }
+
+                    @Override
+                    public KeyStore getTrustStore() {
+                        return ts;
+                    }
+                };
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        }
+    }
+----
+
+== Startup checks
+
+When the application starts, the TLS registry performs some checks to ensure the configuration is correct:
+
+- the key stores and trust stores are accessible
+- the aliases are available and accessible in the key stores and trust stores
+- the certificates are valid
+- the cipher suites and protocols are valid
+- the CRLs are valid
```

Comment:
These checks happen today anyway. We just move them to startup time so we can provide a better error message (better than TLS handshake failed).

The duration is largely dominated by the files loading (> 90% of the time), which you need to do during the server/client configuration anyway. Note that configuring the server/client does not reload the file to avoid paying double (we keep the buffer in memory and configure the server/client directly with the buffer or the KeyStore object).

If it becomes a problem, we can add a flag so skip the validation (and enjoy TLS handshake exceptions)

Reply:
👍🏼